hack/tools/fetch-images/main.go

```diff
@@ -0,0 +1,290 @@
+	"text/template"
+
+	"github.com/d2iq-labs/helm-list-images/pkg"
+	"github.com/d2iq-labs/helm-list-images/pkg/k8s"
+	yamlv2 "gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const (
+	createImagesCMD          = "create-images"
+	defaultHelmAddonFilename = "values-template.yaml"
+)
+
+type ChartInfo struct {
+	repo         string
+	name         string
+	valuesFile   string
+	stringValues []string
+}
+
+func main() {
+	args := os.Args
+	var (
+		chartDirectory     string
+		helmChartConfigMap string
+		carenVersion       string
+	)
+	flagSet := flag.NewFlagSet(createImagesCMD, flag.ExitOnError)
+	flagSet.StringVar(&chartDirectory, "chart-directory", "",
+		"path to chart directory for CAREN")
+	flagSet.StringVar(&helmChartConfigMap, "helm-chart-configmap", "",
+		"path to chart directory for CAREN")
+	flagSet.StringVar(&carenVersion, "caren-version", "",
+		"caren version for images override")
+	err := flagSet.Parse(args[1:])
+	if err != nil {
+		fmt.Println("failed to parse args", err.Error())
+		os.Exit(1)
+	}
+	chartDirectory, err = EnsureFullPath(chartDirectory)
+	if err != nil {
+		fmt.Println("failed to get full path for chart", err.Error())
+		os.Exit(1)
+	}
+	helmChartConfigMap, err = EnsureFullPath(helmChartConfigMap)
+	if err != nil {
+		fmt.Println("failed to get full path for helmChartConfigMap", err.Error())
+		os.Exit(1)
+	}
+	if chartDirectory == "" || helmChartConfigMap == "" {
+		fmt.Println("chart-directory helm-chart-configmap must be set")
+		os.Exit(1)
+	}
+	i := &ChartInfo{
+		name: chartDirectory,
+	}
+	if carenVersion != "" {
+		i.stringValues = []string{
+			fmt.Sprintf("image.tag=%s", carenVersion),
+			fmt.Sprintf("helmRepositoryImage.tag=%s", carenVersion),
+		}
+	}
+	images, err := getImagesForChart(i)
+	if err != nil {
+		fmt.Println("failed to get images", err.Error())
+		os.Exit(1)
+	}
+	addonImages, err := getImagesForAddons(helmChartConfigMap, chartDirectory)
+	if err != nil {
+		fmt.Println("failed to get images from addons", err.Error())
+		os.Exit(1)
+	}
+	images = append(images, addonImages...)
+	images = slices.Compact(images)
+	for _, image := range images {
+		fmt.Println(image)
+	}
+}
+
+func EnsureFullPath(filename string) (string, error) {
+	if path.IsAbs(filename) {
+		return filename, nil
+	}
+	wd, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("failed to get wd: %w", err)
+	}
+	fullPath := path.Join(wd, filename)
+	fullPath = path.Clean(fullPath)
+	_, err = os.Stat(fullPath)
+	if err != nil {
+		return "", err
+	}
+	return fullPath, nil
+}
+
+type HelmChartFromConfigMap struct {
+	Version    string `yaml:"ChartVersion"`
+	Repository string `yaml:"RepositoryURL"`
+	ChartName  string `yaml:"ChartName"`
+}
+
+func getImagesForAddons(helmChartConfigMap, carenChartDirectory string) ([]string, error) {
```

Comment:
Currently Calico uses an operator that then pulls additional images.
There is some additional tooling that generates a list of images.

Do you think its the responsibility of this tool to have special handling for certain addons? (just Calico for now)
```
calico_cni_images_for_version() {
  local -r calico_operator_version="${1}"
  local -r calico_versions="$(curl -s https://raw.githubusercontent.com/tigera/operator/"${calico_operator_version}"/config/calico_versions.yml)"
  local -r calico_registry="${2:-docker.io}"
  # TODO Parse the images and versions from https://github.com/tigera/operator/blob/${calico_operator_version}/pkg/components/calico.go; it appears to be a better source, because it has the exact image names.
  local -a calico_images=("quay.io/tigera/operator:${calico_operator_version}")
  calico_images+=("${calico_registry}/calico/cni:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/cni"].version')")
  calico_images+=("${calico_registry}/calico/kube-controllers:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/kube-controllers"].version')")
  calico_images+=("${calico_registry}/calico/node:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/node"].version')")
  calico_images+=("${calico_registry}/calico/apiserver:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/apiserver"].version')")
  calico_images+=("${calico_registry}/calico/pod2daemon-flexvol:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output .components.flexvol.version)")
  calico_images+=("${calico_registry}/calico/typha:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output .components.typha.version)")
  calico_images+=("${calico_registry}/calico/csi:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/csi"].version')")
  calico_images+=("${calico_registry}/calico/node-driver-registrar:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["csi-node-driver-registrar"].version')")
  # The `calico/ctl` image contains the Calico CLI, which can be useful for debugging.
  calico_images+=("${calico_registry}/calico/ctl:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/cni"].version')")

  IMAGES=("${IMAGES[@]}" "${calico_images[@]}")
}
```

Reply:
The helm list images fork that we use here has the capability to add extra images that cannot be discovered automatically

See https://github.com/d2iq-labs/helm-list-images/blob/f30cfc9a97ff64f08badcf12082d1a7d3b78c966/docs/doc/list-images.md?plain=1#L27 for details with an example extra file at https://github.com/d2iq-labs/helm-list-images/blob/f30cfc9a97ff64f08badcf12082d1a7d3b78c966/extra.yaml#L3 (set how you can use helm values that will be templated here).

Reply:
These images already get accounted for by a process we have to collect all of the cluster resource set images. I don't think it's strictly necessary to handle this right now.

Reply:
> These images already get accounted for by a process we have to collect all of the cluster resource set images.

Is that some process in this repo?

Reply:
No, it's in a downstream dependency of this repo.

Reply:
It seems preferable that this project lists all the images required for all addons deployed by this project. We can do it as a follow up though, I don't mind adding the extra images in a follow up PR.

Reply:
sounds good- I couldn’t tell what the original ask was when this comment was written. “responsibility” could be argued either way.

Reply:
Pushed a PR in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/900.